### PR TITLE
Tests for fasta, etc. cli output options

### PR DESCRIPTION
**Pull request by dcolligan**
Issue #742

### CODE REVIEW

File: tests/unit/test_cli.py [hunk 1]
```diff
@@ -211,3 +214,63 @@ def testReferenceBasesListArguments(self):
...
+    def _getRunPrintMethodCalls(self, runner):
+        printCalls = []
+        with mock.patch('__builtin__.print') as printMethod:
+            printMethod.side_effect = \
+                lambda *args, **kwargs: printCalls.append((args, kwargs))
+            runner.run()
+        return printCalls
+
+    def testListReferenceBasesFasta(self):
+        args = self.FakeArgs('fasta')
```
Comment (jeromekelleher) [inline tests/unit/test_cli.py]: Mocks are neat, but I find them hard to follow and (particularly) maintain sometimes. Here we're violating some interface rules by accessing the ``_client`` attribute, and we also need to know the detailed interface of  the ``listReferenceBases`` method. Depending on the internal structures in the class makes it difficult to change things, as we don't know what tests we're going to break, or necessarily how to fix the tests that have been broken afterwards. I hit a lot of this in the last refactoring pass.

I would _much_ rather depend on the external, advertised interface for tests. Here for example, why not use a local simulated backend, and capture the output? We can then compare the fasta output with the truth that we get from the backend.

Comment (dcolligan) [inline tests/unit/test_cli.py]: two reasons:
1) that's more integration test-y than unit test-y ... mocks make it easy to isolate a particular piece of code and not test any other coupled pieces of code (for instance, if the simulated backend breaks, your version of the test would fail)
2) it would be difficult to capture the output without invoking client_dev.py in another process, since run calls directly to print

Comment (jeromekelleher) [inline tests/unit/test_cli.py]: OK, point (1) is well taken, and convincing.  We also need integrationy test for this though--- we should test the end-to-end correctness of FASTA output.

Point (2) isn't true any more though: we now have some rather snazzy machinery for running CLI commands and capturing the output, without forking a subprocess. See ``tests/end_to_end/test_client_json.py`` for examples.

**Comment by dcolligan** [comment]
@jeromekelleher updated with end to end test

File: tests/end_to_end/test_client_json.py [hunk 2]
```diff
@@ -28,17 +26,54 @@ def setUp(self):
...
+    Tests client FASTA output
+    """
+    def captureFastaOutput(self, command, arguments=""):
+        stdout = self.captureCliOutput(command, arguments, "fasta")
+        lines = stdout.split()
+        return lines
+
+    def testListReferenceBases(self):
+        referenceSetIterator = self._client.searchReferenceSets()
+        referenceSet = referenceSetIterator.next()
```
Comment (jeromekelleher) [inline tests/end_to_end/test_client_json.py]: Calling ``.next`` is Python 2 only --- the Python 2/3 way to do it is to call ``next(referenceSetIterator)`` instead.

**Comment by jeromekelleher** [comment]
Thanks @dcolligan, looks good. +1 modulo the  ``next`` thing above.

**Comment by dcolligan** [comment]
@jeromekelleher updated

**Comment by jeromekelleher** [comment]
Thanks @dcolligan, merging.
